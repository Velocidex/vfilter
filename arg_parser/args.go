// Utility functions for extracting and validating inputs to functions
// and plugins.
package arg_parser

import (
	"context"
	"fmt"
	"reflect"

	"github.com/Velocidex/ordereddict"
	"www.velocidex.com/golang/vfilter/types"
	"www.velocidex.com/golang/vfilter/utils"
)

// Extract the content of args into the struct value. Value's members
// should be tagged with the "vfilter" tag.

// This function makes it very easy to extract args into VQL plugins
// or functions. Simply declare an args struct:

// type MyArgs struct {
//    Field string `vfilter:"required,field=field_name"`
// }

// And parse the struct using this function:
// myarg := &MyArgs{}
// err := vfilter.ExtractArgs(scope, args, myarg)

// We will raise an error if a required field is missing or has the
// wrong type of args.

// NOTE: In order for the field to be populated by this function, the
// field must be exported (i.e. name begins with cap) and it must have
// vfilter tags.

// Deprecate this in favor of ExtractArgsWithContext
func ExtractArgs(scope types.Scope, args *ordereddict.Dict, target interface{}) error {
	v := reflect.ValueOf(target)
	if v.Type().Kind() == reflect.Ptr {
		v = v.Elem()
	}

	parser, err := GetParser(v)
	if err != nil {
		return err
	}

	return parser.Parse(context.Background(), scope, args, v)
}

func ExtractArgsWithContext(
	ctx context.Context, scope types.Scope, args *ordereddict.Dict, target interface{}) error {
	v := reflect.ValueOf(target)
	if v.Type().Kind() == reflect.Ptr {
		v = v.Elem()
	}

	parser, err := GetParser(v)
	if err != nil {
		return err
	}

	return parser.Parse(ctx, scope, args, v)
}

// Try to retrieve an arg name from the Dict of args. Coerce the arg
// into something resembling a list of strings.
func _ExtractStringArray(
	ctx context.Context, scope types.Scope, arg types.Any) ([]string, bool) {
	var result []string

	// Handle potentially lazy args.
	lazy_arg, ok := arg.(types.LazyExpr)
	if ok {
		arg = lazy_arg.Reduce(ctx)
	}

	slice := reflect.ValueOf(arg)

	// A slice of strings.
	if slice.Type().Kind() == reflect.Slice {
		for i := 0; i < slice.Len(); i++ {
			value := slice.Index(i).Interface()
			item, ok := utils.ToString(value)
			if ok {
				result = append(result, item)
				continue
			}

			// If is a dict with only one member just use
			// that one.
			members := scope.GetMembers(value)
			if len(members) == 1 {
				member, ok := scope.Associative(value, members[0])
				if ok {
					item, ok := utils.ToString(member)
					if ok {
						result = append(result, item)
					}
				}
			}

			// Represent the value as a string.
			result = append(result, fmt.Sprintf("%v", value))
		}
		return result, true
	}

	// A single string just expands into a list of length 1.
	item, ok := utils.ToString(slice.Interface())
	if !ok {
		// If it has no StringProtocol fall back to golang
		// default.
		item = fmt.Sprintf("%v", slice.Interface())
	}
	result = append(result, item)
	return result, true
}

// Convert a type to a stored query
func ToStoredQuery(ctx context.Context, arg types.Any) types.StoredQuery {
	switch t := arg.(type) {
	case types.LazyExpr:
		return ToStoredQuery(ctx, t.Reduce(ctx))

	case types.StoredQuery:
		return t
	default:
		return &storedQueryWrapper{arg}
	}
}

type storedQueryWrapper struct {
	value types.Any
}

func (self *storedQueryWrapper) Eval(ctx context.Context, scope types.Scope) <-chan types.Row {
	output_chan := make(chan types.Row)
	go func() {
		defer close(output_chan)

		slice := reflect.ValueOf(self.value)
		if slice.Type().Kind() == reflect.Slice {
			for i := 0; i < slice.Len(); i++ {
				value := slice.Index(i).Interface()
				if !types.IsNullObject(value) {
					select {
					case <-ctx.Done():
						return
					case output_chan <- self.toRow(scope, value):
					}
				}
			}
		} else {
			row_value := self.toRow(scope, self.value)
			if !types.IsNullObject(row_value) {
				select {
				case <-ctx.Done():
					return
				case output_chan <- row_value:
				}
			}
		}

	}()
	return output_chan
}

func (self *storedQueryWrapper) toRow(scope types.Scope, value types.Any) types.Row {
	if types.IsNullObject(value) {
		return types.Null{}
	}

	members := scope.GetMembers(value)
	if len(members) > 0 {
		return value
	}

	return ordereddict.NewDict().Set("_value", value)
}

// Wrap an arg in a LazyExpr for plugins that want to receive a
// LazyExpr.
func ToLazyExpr(scope types.Scope, arg types.Any) types.LazyExpr {
	switch t := arg.(type) {
	case types.LazyExpr:
		return t

	case types.StoredQuery:
		return &storedQueryWrapperLazyExpression{query: t}
	default:
		return &lazyExpressionWrapper{arg}
	}
}

// Wrap a Stored Query with a LazyExpr interface. Callers will receive
// the Stored Query when reducing us.
type storedQueryWrapperLazyExpression struct {
	query types.StoredQuery
}

func (self *storedQueryWrapperLazyExpression) ReduceWithScope(
	ctx context.Context, scope types.Scope) types.Any {
	return types.Materialize(ctx, scope, self.query)
}

func (self *storedQueryWrapperLazyExpression) Reduce(ctx context.Context) types.Any {
	return self.query
}

type lazyExpressionWrapper struct {
	value types.Any
}

func (self *lazyExpressionWrapper) ReduceWithScope(ctx context.Context, scope types.Scope) types.Any {
	return self.value
}

func (self *lazyExpressionWrapper) Reduce(ctx context.Context) types.Any {
	return self.value
}

type StringProtocol interface {
	ToString() string
}
