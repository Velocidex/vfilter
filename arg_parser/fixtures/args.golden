{
  "000/000 Parse basic types: SELECT parse(r=1, int=1, string='hello') FROM scope()": [
    {
      "parse(r=1, int=1, string='hello')": {
        "int": 1,
        "string": "hello"
      }
    }
  ],
  "001/000 Parse basic types: LET X = 5": null,
  "001/001 Parse basic types: SELECT parse(r=1, int=X) FROM scope()": [
    {
      "parse(r=1, int=X)": {
        "int": 5
      }
    }
  ],
  "002/000 Parse basic types with param: LET Foo(X) = 1 + X": null,
  "002/001 Parse basic types with param: SELECT parse(r=1, int=Foo(X=2)) FROM scope()": [
    {
      "parse(r=1, int=Foo(X=2))": {
        "int": 3
      }
    }
  ],
  "003/000 Passing Stored query to int field: LET Foo = SELECT 1 FROM scope()": null,
  "003/001 Passing Stored query to int field: SELECT parse(r=1, int=Foo) FROM scope()": [
    {
      "parse(r=1, int=Foo)": {
        "ParseError": "Field int should be an int."
      }
    }
  ],
  "004/000 Passing string to int field: LET Foo = \"Hello\"": null,
  "004/001 Passing string to int field: SELECT parse(r=1, int=Foo) FROM scope()": [
    {
      "parse(r=1, int=Foo)": {
        "ParseError": "Field int should be an int."
      }
    }
  ],
  "005/000 String Array: SELECT parse(r=1, string_array=[\"X\", \"Y\"]) FROM scope()": [
    {
      "parse(r=1, string_array=[\"X\", \"Y\"])": {
        "string_array": [
          "X",
          "Y"
        ]
      }
    }
  ],
  "006/000 String Array with single field: SELECT parse(r=1, string_array=\"Hello\") FROM scope()": [
    {
      "parse(r=1, string_array=\"Hello\")": {
        "string_array": [
          "Hello"
        ]
      }
    }
  ],
  "007/000 String Array getting int array stringifies it: SELECT parse(r=1, string_array=[1, ]) FROM scope()": [
    {
      "parse(r=1, string_array=[1, ])": {
        "string_array": [
          "1"
        ]
      }
    }
  ],
  "008/000 String Array getting int stringifies it: SELECT parse(r=1, string_array=1) FROM scope()": [
    {
      "parse(r=1, string_array=1)": {
        "string_array": [
          "1"
        ]
      }
    }
  ],
  "009/000 String Array with single field: LET Foo = \"Hello\"": null,
  "009/001 String Array with single field: SELECT parse(r=1, string_array=Foo) FROM scope()": [
    {
      "parse(r=1, string_array=Foo)": {
        "string_array": [
          "Hello"
        ]
      }
    }
  ],
  "010/000 String Array with stored query expanding a row: LET Foo = SELECT \"Hello\" AS X FROM scope()": null,
  "010/001 String Array with stored query expanding a row: SELECT parse(r=1, string_array=Foo.X) FROM scope()": [
    {
      "parse(r=1, string_array=Foo.X)": {
        "string_array": [
          "Hello"
        ]
      }
    }
  ],
  "011/000 String Array with stored query expanding a row of ints: LET Foo = SELECT 1 AS X FROM scope()": null,
  "011/001 String Array with stored query expanding a row of ints: SELECT parse(r=1, string_array=Foo.X) FROM scope()": [
    {
      "parse(r=1, string_array=Foo.X)": {
        "string_array": [
          "1"
        ]
      }
    }
  ],
  "012/000 Lazy expressions: LET lazy_expr = 1": null,
  "012/001 Lazy expressions: SELECT parse(r=1, lazy=lazy_expr) FROM scope()": [
    {
      "parse(r=1, lazy=lazy_expr)": {
        "Lazy type": "*vfilter.LazyExprImpl",
        "Lazy Reduced Type": "int64",
        "Lazy Reduced": 1
      }
    }
  ],
  "013/000 Lazy expressions with parameters: LET lazy_expr(X) = X + 1": null,
  "013/001 Lazy expressions with parameters: SELECT parse(r=1, lazy=lazy_expr(X=1)) FROM scope()": [
    {
      "parse(r=1, lazy=lazy_expr(X=1))": {
        "Lazy type": "*vfilter.LazyExprImpl",
        "Lazy Reduced Type": "int64",
        "Lazy Reduced": 2
      }
    }
  ],
  "014/000 Lazy expressions of stored query: LET query = SELECT 1 FROM scope()": null,
  "014/001 Lazy expressions of stored query: SELECT parse(r=1, lazy=query) FROM scope()": [
    {
      "parse(r=1, lazy=query)": {
        "Lazy type": "*vfilter.LazyExprImpl",
        "Lazy Reduced Type": "*vfilter._StoredQuery",
        "Lazy Reduced": {},
        "Lazy Reduced stored query": [
          {
            "1": 1
          }
        ]
      }
    }
  ],
  "015/000 Lazy expressions of stored query with parameters: LET X = 5": null,
  "015/001 Lazy expressions of stored query with parameters: LET query(X) = SELECT X FROM scope()": null,
  "015/002 Lazy expressions of stored query with parameters: SELECT parse(r=1, lazy=query(X=2)) FROM scope()": [
    {
      "parse(r=1, lazy=query(X=2))": {
        "Lazy type": "*vfilter.LazyExprImpl",
        "Lazy Reduced Type": "*vfilter.StoredQueryCallSite",
        "Lazy Reduced": {},
        "Lazy Reduced stored query": [
          {
            "X": 2
          }
        ]
      }
    }
  ],
  "016/000 Stored query: LET query = SELECT 1 FROM scope()": null,
  "016/001 Stored query: SELECT parse(r=1, query=query) FROM scope()": [
    {
      "parse(r=1, query=query)": {
        "StoredQuery Materialized": [
          {
            "1": 1
          }
        ]
      }
    }
  ],
  "017/000 Stored query with parameters: LET X = 5": null,
  "017/001 Stored query with parameters: LET query(X) = SELECT X FROM scope()": null,
  "017/002 Stored query with parameters: SELECT parse(r=1, query=query(X=2)) FROM scope()": [
    {
      "parse(r=1, query=query(X=2))": {
        "StoredQuery Materialized": [
          {
            "X": 2
          }
        ]
      }
    }
  ],
  "018/000 Stored query given a constant: SELECT parse(r=1, query=\"hello\") FROM scope()": [
    {
      "parse(r=1, query=\"hello\")": {
        "StoredQuery Materialized": [
          {
            "_value": "hello"
          }
        ]
      }
    }
  ],
  "019/000 Stored query given a dict: SELECT parse(r=1, query=dict(X=\"hello\")) FROM scope()": [
    {
      "parse(r=1, query=dict(X=\"hello\"))": {
        "StoredQuery Materialized": [
          {
            "X": "hello"
          }
        ]
      }
    }
  ],
  "020/000 Stored query given an expression: LET X = 1": null,
  "020/001 Stored query given an expression: SELECT parse(r=1, query=X) FROM scope()": [
    {
      "parse(r=1, query=X)": {
        "StoredQuery Materialized": [
          {
            "_value": 1
          }
        ]
      }
    }
  ],
  "021/000 Any type: LET X = 1": null,
  "021/001 Any type: SELECT parse(r=1, any=X) FROM scope()": [
    {
      "parse(r=1, any=X)": {
        "any": 1,
        "any type": "int64"
      }
    }
  ],
  "022/000 Any type: LET Foo(X) = X + 1": null,
  "022/001 Any type: SELECT parse(r=1, any=Foo(X=1)) FROM scope()": [
    {
      "parse(r=1, any=Foo(X=1))": {
        "any": 2,
        "any type": "int64"
      }
    }
  ],
  "023/000 Any type: LET query = SELECT 1 FROM scope()": null,
  "023/001 Any type: SELECT parse(r=1, any=query) FROM scope()": [
    {
      "parse(r=1, any=query)": {
        "any": {},
        "any type": "*vfilter._StoredQuery",
        "Any stored query": [
          {
            "1": 1
          }
        ]
      }
    }
  ],
  "024/000 Dict type: SELECT parse(r=1, dict=dict(Foo=1)) FROM scope()": [
    {
      "parse(r=1, dict=dict(Foo=1))": {
        "dict": {
          "Foo": 1
        }
      }
    }
  ],
  "025/000 Unexpected args: SELECT parse(r=1, int=1, foobar=2) FROM scope()": [
    {
      "parse(r=1, int=1, foobar=2)": {
        "ParseError": "Unexpected arg foobar"
      }
    }
  ],
  "026/000 Required args: SELECT parse() FROM scope()": [
    {
      "parse()": {
        "ParseError": "Field r is required"
      }
    }
  ]
}